backend/ranking.js

```diff
@@ -3,10 +3,184 @@ import { Follow } from "./models/follow.js";
 import { Like } from "./models/like.js";
 import { Dislike } from "./models/dislike.js";
 import { compareStocksByPoints } from "./utils.js";
+import { User } from "./models/user.js";
+import { Friend } from "./models/friend.js";
 
 const FOLLOW_POINTS = 25;
 const LIKE_POINTS = 20;
 const DISLIKE_POINTS = -20;
+const FOLLOW_FRIEND_POINTS = 10;
+const LIKE_FRIEND_POINTS = 10;
+const DISLIKE_FRIEND_POINTS = -10;
+
+export async function RankingV3(user) {
+    try {
+        const [stocksInDatabase, follows, likes, dislikes, connections] = await Promise.all([
+            Stock.findAll(),
+            Follow.findAll({
+                where: { UserId: user.id },
+                include: [{ model: Stock }]
+            }),
+            Like.findAll({
+                where: { UserId: user.id },
+                include: [{ model: Stock }]
+            }),
+            Dislike.findAll({
+                where: { UserId: user.id },
+                include: [{ model: Stock }]
+            }),
+            Friend.findAll({
+                where: { UserId1: user.id },
+                include: [
+                    { model: User, as: "user1" },
+                    { model: User, as: "user2" }
+                ]
+            })
+        ]);
+
+        if (stocksInDatabase.length < 10) {
+            return {
+                status: 422,
+                error: "You need at least 10 stocks in the database to access ranking."
+            };
+        }
+
+        const followsMap = follows.reduce((accum, current) => {
+            accum[current.Stock.ticker] = true;
+            return accum;
+        }, {});
+
+        const likesMap = likes.reduce((accum, current) => {
+            accum[current.Stock.ticker] = true;
+            return accum;
+        }, {});
+
+        const dislikesMap = dislikes.reduce((accum, current) => {
+            accum[current.Stock.ticker] = true;
+            return accum;
+        }, {});
+
+        const friendIDs = connections.map((connection) => connection.user2.id);
+
+        const stocksThatFriendsFollow = await Promise.all(
+            friendIDs.map(async (friendID) => {
+                const friendFollows = await Follow.findAll({
+                    where: { UserId: friendID },
+                    include: [{ model: Stock }]
+                });
+                const stocksTheyFollow = friendFollows.map((friendFollow) => {
+                    return friendFollow.Stock.dataValues;
```

Comment:
To query the follows at the top you use:
```
            Follow.findAll({
                where: { UserId: user.id },
                include: [{ model: Stock }]
            }),
```

Which is the same query used here. I don't understand why in this case further processing is needed (`friendFollow.Stock.dataValues`).  
Wouldn't it work if only friendFollows is returned here?